components/supervisor/pkg/supervisor/top.go

```diff
@@ -10,17 +10,73 @@ import (
 	"os"
 	"strconv"
 	"strings"
+	"sync"
 	"time"
 
 	"golang.org/x/xerrors"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials/insecure"
 
 	linuxproc "github.com/c9s/goprocinfo/linux"
+	"github.com/gitpod-io/gitpod/common-go/log"
 	"github.com/gitpod-io/gitpod/supervisor/api"
 	daemonapi "github.com/gitpod-io/gitpod/ws-daemon/api"
 )
 
+type TopService struct {
+	data      *api.ResourcesStatusResponse
+	ready     chan struct{}
+	readyOnce sync.Once
+	Top       func(ctx context.Context) (*api.ResourcesStatusResponse, error)
+}
+
+func NewTopService(
+	top func(ctx context.Context) (*api.ResourcesStatusResponse, error),
+) *TopService {
+	log.Debug("gitpod top service: initialized")
+	return &TopService{
+		Top: top,
```

Comment:
(nit) fyi you could do here like `top: Top` and in test do `topSerivce.top = func() {}`

Reply:
ok, it would be nicer if in `supervisor.go` I don't have to pass `Top` as func in the constructor but in this case the constructor I guess I don't need it anymore and only do what you suggest in the tests `topSerivce.top = func() {}`

Reply:
I think you can keep it, since you encapsulated how top is implemented?

Reply:
I must have misinterpreted your suggestion, if I do `top: Top` it will pick up the globally exported `Top` func. 🤔

Reply:
I meant:
```go
func NewTopService() *TopService {
	log.Debug("gitpod top service: initialized")
	return &TopService{
		top: Top,
	}
}
```